lib/web_ui/lib/src/engine/canvas_pool.dart

```diff
@@ -747,10 +748,124 @@ class ContextStateHandle {
+      fillStyle = colorString;
+      strokeStyle = colorString;
+    } else {
+      fillStyle = '';
+      strokeStyle = '';
+    }
+
+    final ui.MaskFilter maskFilter = paint?.maskFilter;
+    if (!_renderMaskFilterForWebkit) {
+      if (_currentFilter != maskFilter) {
+        _currentFilter = maskFilter;
+        context.filter = _maskFilterToCanvasFilter(maskFilter);
+      }
+    } else {
+      // WebKit does not support the `filter` property. Instead we apply a
+      // shadow to the shape of the same color as the paint and the same blur
+      // as the mask filter.
+      //
+      // Note that on WebKit the cached value of _currentFilter is not useful.
+      // Instead we destructure it into the shadow properties and cache those.
+      if (maskFilter != null) {
+        context.save();
+        context.shadowBlur = convertSigmaToRadius(maskFilter.webOnlySigma);
+        if (paint?.color != null) {
+          // Shadow color must be fully opaque.
+          context.shadowColor = colorToCssString(paint.color.withAlpha(255));
+        } else {
+          context.shadowColor = colorToCssString(const ui.Color(0xFF000000));
+        }
+
+        // On the web a shadow must always be painted together with the shape
+        // that casts it. In order to paint just the shadow, we offset the shape
+        // by a large enough value that it moved outside the canvas bounds, then
+        // offset the shadow in the opposite direction such that it lands exactly
+        // where the shape is.
+        const double kOutsideTheBoundsOffset = 50000;
+
+        context.translate(-kOutsideTheBoundsOffset, 0);
+
+        // Shadow offset is not affected by the current canvas context transform.
+        // We have to apply the transform ourselves. To do that we transform the
+        // tip of the vector from the shape to the shadow, then we transform the
+        // origin (0, 0). The desired shadow offset is the difference between the
+        // two. In vector notation, this is:
+        //
+        // transformedShadowDelta = M*shadowDelta - M*origin.
+        final Float32List tempVector = Float32List(2);
+        tempVector[0] = kOutsideTheBoundsOffset * window.devicePixelRatio;
+        _canvasPool.currentTransform.transform2(tempVector);
+        double shadowOffsetX = tempVector[0];
+        double shadowOffsetY = tempVector[1];
+
+        tempVector[0] = tempVector[1] = 0;
+        _canvasPool.currentTransform.transform2(tempVector);
+        context.shadowOffsetX = shadowOffsetX - tempVector[0];
+        context.shadowOffsetY = shadowOffsetY - tempVector[1];
+      }
+    }
+  }
+
+  /// Removes paint properties on the current canvas used by the last draw
+  /// command.
+  ///
+  /// Not all properties are cleared. Properties that are set by all paint
+  /// commands prior to painting do not need to be cleared.
+  ///
+  /// Must be called after calling [setUpPaint].
+  void tearDownPaint() {
```

Comment:
Does this get removed dart2js release? Maybe place all tearDownPaint() calls inside asserts?

Reply:
It should not be removed. Otherwise the context won't be restored.

Reply:
Got it. likely very small will be inlined

Reply:
Would be nice if we had access to dart2js inlining annotations.